design/service-apis.md

```diff
@@ -0,0 +1,249 @@
+
+## Goals
+
+- Manage Contour and dependent resources using Service APIs.
+
+## Non Goals
+
+- Support application-level Service APIs, i.e. xRoute. These APIs will be supported by Contour directly according to
+  Issue [2809][2].
+
+## Definitions
+
+### Service APIs
+
+Service APIs is an open source project managed by the Kubernetes SIG-NETWORK community. The project's goal is to evolve
+service networking APIs within the Kubernetes ecosystem. Service APIs provide interfaces to expose Kubernetes
+applications- Services, Ingress, and more.
+
+## Background
+
+Contour Operator manages Contour using the `contours.operator.projectcontour.io` Custom Resource Definition (CRD). The
+Contour Custom Resource (CR) provides the user interface for managing an instance of Contour and is handled by the
+Kubernetes API just like built-in objects. Contour Operator monitors the Contour CR and provides status on whether the
+actual state matches the desired state of the object. Contour Operator will continue to support the Contour CRD. Adding
+support for Service APIs will provide users an alternative approach to managing Contour.
+
+## High-Level Design
+
+Contour Operator will manage Contour using the `gatewayclasses.networking.x-k8s.io` and `gateways.networking.x-k8s.io`
+CRDs. The combination of GatewayClass and Gateway CRs will provide the user interface for managing an instance of
+Contour. Service APIs provides extension points to allow for custom resources to be linked at various layers of the API.
+This provides more granular customization at the appropriate places within the API structure. These extension points
+will be leveraged to expose Contour-specific configuration.
+
+## Detailed Design
+
+### GatewayClass
+
+GatewayClass is a cluster-scoped resource that defines a set of Gateways sharing a common behavior. A GatewayClass
+with `controller: projectcontour.io/contour-operator` will be managed by Contour Operator. There MUST be at least one
+GatewayClass defined in order to be able to have a functional Gateway. This is similar to IngressClass for Ingress and
+StorageClass for PersistentVolumes. The following is an example of a GatewayClass managed by Contour Operator:
+
+```yaml
+kind: GatewayClass
+apiVersion: networking.x-k8s.io/v1alpha1
+metadata:
+  name: example-gatewayclass
+spec:
+  controller: projectcontour.io/contour-operator
+  parametersRef:
+    name: example-contour-config
+    group: operator.projectcontour.io
+    kind: ContourConfig
+```
+
+The `controller` field is a domain/path string that indicates the controller responsible for managing Gateways of this
+class. The `projectcontour.io/contour-operator` string will be used to specify Contour Operator as the controller
+responsible for handling the GatewayClass. This field is not mutable and cannot be empty. Versioning can be accomplished
+by encoding the operator's version into the path. For example:
+
+```
+projectcontour.io/contour-operator/v1   // Use version 1
```

Comment:
What version does this relate to? The verison of Contour? The version of the operator?

Reply:
The field can be used to specify an operator version to use for gateways of this `GatewayClass`. Since the operator version is synchronized with the contour version, this field would also represent the Contour version.

Reply:
So today you'd have this: `projectcontour.io/contour-operator/v1.11.0`? Then you'd upgrade the operator to say v1.12.0 and you'd need to then change the gateway to be `projectcontour.io/contour-operator/v1.12.0` to pick up the update?

What happens with the older versions, would they be orphaned then since the operator only works on `latest`?

Reply:
> Then you'd upgrade the operator to say v1.12.0

@stevesloka this is the workflow that I envision:
- The infrastructure provider upgrades the operator image to v1.12.0, triggering an upgrade of the operator. The Contour CRD is updated if API changes are included in v1.12.0.
- The status of `projectcontour.io/contour-operator/v1.11.0` GatewayClasses are updated to "Waiting" and associated Gateways are updated to "NoSuchGatewayClass". The details for accomplishing this step is TBD and I welcome any input. Should Contour be responsible for updating GatewayClass/Gateway/xRoute status when it's in this "detached" state?
- The operator will no longer manage Gateways of `projectcontour.io/contour-operator/v1.11.0` GatewayClasses. These Gateways will continue to function and traffic will continue to be proxied. Contour will continue running v1.11.0 and Envoy v1.16.2.
- Optionally, the infrastructure provider updates the Contour resource referenced by the GatewayClass, or creates and associates a new Contour to the GatewayClass.
- The infrastructure provider updates Gateways by setting GatewayClasses to `projectcontour.io/contour-operator/v1.12.0`.  This triggers the operator to update Contour to v1.12.0 and optionally update the configuration. GatewayClass/Gateway/xRoute status conditions are updated accordingly.

__Note:__ GatewayClasses of `projectcontour.io/contour-operator` are equivalent to running the latest operator release. In this case, Gateways will be automatically upgraded when the operator is upgraded. `projectcontour.io/contour-operator` will be the only supported value for the initial support of Service APIs.

Thoughts?

Reply:
It feels like there's too much logic applied with convention to the operator. We might need a version field in the spec to define what version of Contour you'd like to run for a specific instance of Contour that's managed by the operator. Even if that field could be set to `latest` if folks want truly the latest version or if they want to manage when the upgrade gets applied to their instances. A good use-case might be to upgrade the external version of Contour before upgrading the internal version. 

> The status of projectcontour.io/contour-operator/v1.11.0 GatewayClasses are updated to "Waiting" and associated Gateways are updated to "NoSuchGatewayClass". 

How did this happen? Did the operator change the GatewayClass somehow? It seems like if you upgraded the instance of the operator, and it wasn't programmed to look at a specific class, it would just ignore? 

> Optionally, the infrastructure provider updates the Contour resource referenced by the GatewayClass, or creates and associates a new Contour to the GatewayClass.

We need to be careful about how users need to manage their external services tied to Envoy, since in many cases, they are associated with DNS, etc.